File: ios/Classes/ShareHarborPlugin.swift
import Flutter
import UIKit

public class ShareHarborPlugin: NSObject, FlutterPlugin, ShareHarborHostApi {

    private var storageCore: NativeStorageCore?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ShareHarborPlugin()
        instance.storageCore = NativeStorageCore()
        ShareHarborHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }

    public func getPendingDeliveries(completion: @escaping (Result<[NativeDelivery], Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        completion(.success(storageCore.getPendingDeliveries()))
    }

    public func claimDelivery(deliveryId: String, leaseDurationSeconds: Int64, completion: @escaping (Result<NativeClaim, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        do {
            let claim = try storageCore.claimDelivery(deliveryId: deliveryId, leaseDurationSeconds: leaseDurationSeconds)
            completion(.success(claim))
        } catch {
            completion(.failure(error))
        }
    }

    public func claimNextDelivery(leaseDurationSeconds: Int64, completion: @escaping (Result<NativeClaim?, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        do {
            let claim = try storageCore.claimNextDelivery(leaseDurationSeconds: leaseDurationSeconds)
            completion(.success(claim))
        } catch {
            completion(.failure(error))
        }
    }

    public func acknowledgeClaim(claimId: String, deliveryId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        do {
            try storageCore.acknowledgeClaim(claimId: claimId, deliveryId: deliveryId)
            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    public func releaseClaim(claimId: String, deliveryId: String, reason: String?, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        do {
            try storageCore.releaseClaim(claimId: claimId, deliveryId: deliveryId, reason: reason)
            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    public func retryDelivery(deliveryId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        storageCore?.retryDelivery(deliveryId: deliveryId)
        completion(.success(()))
    }

    public func inspectInbox(completion: @escaping (Result<NativeHealth, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        completion(.success(storageCore.inspectInbox()))
    }

    public func cleanupInbox(maxAgeSeconds: Int64, completion: @escaping (Result<NativeCleanupResult, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        completion(.success(storageCore.cleanupInbox(maxAgeSeconds: maxAgeSeconds)))
    }

    public func getPayloadPath(deliveryId: String, itemId: String, completion: @escaping (Result<String, Error>) -> Void) {
        guard let storageCore = storageCore else {
            completion(.failure(NSError(domain: "ShareHarbor", code: 1, userInfo: [NSLocalizedDescriptionKey: "Storage core uninitialized"])))
            return
        }
        do {
            let path = try storageCore.getPayloadPath(deliveryId: deliveryId, itemId: itemId)
            completion(.success(path))
        } catch {
            completion(.failure(error))
        }
    }
}

File: android/src/main/kotlin/dev/raean/share_harbor/ShareHarborPlugin.kt
package dev.raean.share_harbor

import dev.raean.share_harbor.core.NativeStorageCore
import dev.raean.share_harbor.generated.NativeClaim
import dev.raean.share_harbor.generated.NativeCleanupResult
import dev.raean.share_harbor.generated.NativeDelivery
import dev.raean.share_harbor.generated.NativeHealth
import dev.raean.share_harbor.generated.ShareHarborHostApi
import io.flutter.embedding.engine.plugins.FlutterPlugin

class ShareHarborPlugin : FlutterPlugin, ShareHarborHostApi {

    private var storageCore: NativeStorageCore? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val context = flutterPluginBinding.applicationContext
        storageCore = NativeStorageCore(context)
        ShareHarborHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ShareHarborHostApi.setUp(binding.binaryMessenger, null)
        storageCore = null
    }

    override fun getPendingDeliveries(): List<NativeDelivery> {
        return storageCore?.getPendingDeliveries() ?: emptyList()
    }

    override fun claimDelivery(deliveryId: String, leaseDurationSeconds: Long): NativeClaim {
        return storageCore?.claimDelivery(deliveryId, leaseDurationSeconds)
            ?: throw IllegalStateException("Storage core not initialized")
    }

    override fun claimNextDelivery(leaseDurationSeconds: Long): NativeClaim? {
        return storageCore?.claimNextDelivery(leaseDurationSeconds)
    }

    override fun acknowledgeClaim(claimId: String, deliveryId: String) {
        storageCore?.acknowledgeClaim(claimId, deliveryId)
    }

    override fun releaseClaim(claimId: String, deliveryId: String, reason: String?) {
        storageCore?.releaseClaim(claimId, deliveryId, reason)
    }

    override fun retryDelivery(deliveryId: String) {
        storageCore?.retryDelivery(deliveryId)
    }

    override fun inspectInbox(): NativeHealth {
        return storageCore?.inspectInbox()
            ?: NativeHealth(0, 0, 0, 0, listOf("Not initialized"))
    }

    override fun cleanupInbox(maxAgeSeconds: Long): NativeCleanupResult {
        return storageCore?.cleanupInbox(maxAgeSeconds)
            ?: NativeCleanupResult(0, 0)
    }

    override fun getPayloadPath(deliveryId: String, itemId: String): String {
        return storageCore?.getPayloadPath(deliveryId, itemId) ?: ""
    }
}
